File: src/main/java/com/qsoftware/bubbles/graphics/FilterApplier.java
package com.qsoftware.bubbles.graphics;

import com.jhlabs.image.*;
import org.jdesktop.swingx.image.AbstractFilter;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.awt.image.ImageObserver;
import java.util.ArrayList;

public class FilterApplier {
    private final Graphics2D graphics;
    private final BufferedImage buffer;
    private final ImageObserver observer;
    private ArrayList<Object> filters;

    public FilterApplier(Dimension size, ImageObserver observer) {
        size.width = Math.max(size.width, 1);
        size.height = Math.max(size.height, 1);

        BufferedImage buffer = new BufferedImage(size.width, size.height, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g2d = buffer.createGraphics();

        this.observer = observer;
        this.graphics = g2d;
        this.buffer = buffer;
    }

    public Graphics2D getGraphics() {
        return this.graphics;
    }

    public Rectangle getBounds() {
        for (Object filter : filters) {
            if (filter instanceof GaussianFilter) {
                float radius = ((GaussianFilter) filter).getRadius();
                return new Rectangle(
                        (int) -Math.round(Math.ceil(radius)), (int) -Math.round(Math.ceil(radius)),
                        (int) Math.round(Math.ceil(buffer.getWidth() + radius * 2)), (int) Math.round(Math.ceil(buffer.getHeight() + radius * 2)));
            }
        }
        return new Rectangle(0, 0, buffer.getWidth(), buffer.getHeight());
    }

    public Rectangle getBounds(int x, int y, int w, int h) {
        return new Rectangle(x, y, w, h);
    }

    public Rectangle getBounds(Rectangle rect) {
        return getBounds(rect.x, rect.y, rect.width, rect.height);
    }

    public void setFilters(ArrayList<Object> filters) {
        this.filters = filters;
    }

    public BufferedImage done() {
        Rectangle bounds = getBounds();

        BufferedImage outputBuffer = new BufferedImage(bounds.width, bounds.height, BufferedImage.TYPE_INT_ARGB);
        outputBuffer.createGraphics().drawImage(buffer, 0, 0, observer);

        for (Object filter : filters) {
            if (filter instanceof ConvolveFilter) {
                ((ConvolveFilter) filter).filter(outputBuffer, outputBuffer);
            } else if (filter instanceof PointFilter) {
                ((PointFilter) filter).filter(outputBuffer, outputBuffer);
            } else if (filter instanceof TransformFilter) {
                ((TransformFilter) filter).filter(outputBuffer, outputBuffer);
            } else if (filter instanceof BinaryFilter) {
                ((BinaryFilter) filter).filter(outputBuffer, outputBuffer);
            } else if (filter instanceof WholeImageFilter) {
                ((WholeImageFilter) filter).filter(outputBuffer, outputBuffer);
            } else if (filter instanceof ShadowFilter) {
                ((ShadowFilter) filter).filter(outputBuffer, outputBuffer);
            } else if (filter instanceof ShatterFilter) {
                ((ShatterFilter) filter).filter(outputBuffer, outputBuffer);
            } else if (filter instanceof BoxBlurFilter) {
                ((BoxBlurFilter) filter).filter(outputBuffer, outputBuffer);
            } else if (filter instanceof AbstractFilter) {
                ((AbstractFilter) filter).filter(outputBuffer, outputBuffer);
            } else {
                throw new NullPointerException();
            }
        }
        return outputBuffer;
    }
}

File: src/main/kotlin/com/qtech/bubbles/graphics/FilterApplier.kt
package com.qtech.bubbles.graphics

import com.jhlabs.image.*
import org.jdesktop.swingx.image.AbstractFilter
import java.awt.Dimension
import java.awt.Rectangle
import java.awt.image.BufferedImage
import java.awt.image.ImageObserver
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.roundToLong

class FilterApplier(size: Dimension, observer: ImageObserver) {
    val graphics: GraphicsProcessor
    private val buffer: BufferedImage
    private val observer: ImageObserver
    private var filters: ArrayList<Any>? = null
    val bounds: Rectangle
        get() {
            for (filter in filters!!) {
                if (filter is GaussianFilter) {
                    val radius = filter.radius
                    return Rectangle(
                        (-ceil(radius.toDouble()).roundToLong()).toInt(), (-ceil(radius.toDouble()).roundToLong()).toInt(),
                        ceil((buffer.width + radius * 2).toDouble()).roundToLong().toInt(), ceil((buffer.height + radius * 2).toDouble()).roundToLong().toInt()
                    )
                }
            }
            return Rectangle(0, 0, buffer.width, buffer.height)
        }

    fun getBounds(x: Int, y: Int, w: Int, h: Int): Rectangle {
        return Rectangle(x, y, w, h)
    }

    fun getBounds(rect: Rectangle): Rectangle {
        return getBounds(rect.x, rect.y, rect.width, rect.height)
    }

    fun setFilters(filters: ArrayList<Any>?) {
        this.filters = filters
    }

    fun done(): BufferedImage {
        val bounds = bounds
        val outputBuffer = BufferedImage(bounds.width, bounds.height, BufferedImage.TYPE_INT_ARGB)
        outputBuffer.createGraphics().drawImage(buffer, 0, 0, observer)
        for (filter in filters!!) {
            when (filter) {
                is ConvolveFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                is PointFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                is TransformFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                is BinaryFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                is WholeImageFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                is ShadowFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                is ShatterFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                is BoxBlurFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                is AbstractFilter -> {
                    filter.filter(outputBuffer, outputBuffer)
                }
                else -> {
                    throw NullPointerException()
                }
            }
        }
        return outputBuffer
    }

    init {
        size.width = max(size.width, 1)
        size.height = max(size.height, 1)
        val buffer = BufferedImage(size.width, size.height, BufferedImage.TYPE_INT_ARGB)
        val g2d = GraphicsProcessor(buffer.createGraphics())
        this.observer = observer
        graphics = g2d
        this.buffer = buffer
    }
}